pkg/epp/scheduling/scheduler.go

```diff
@@ -20,141 +20,197 @@ package scheduling
+
+	pods, err := s.runFilterPlugins(sCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := s.runScorerPlugins(sCtx, pods); err != nil {
+		return nil, err
+	}
+
+	before := time.Now()
+	res, err := s.picker.Pick(sCtx, pods)
+	metrics.RecordSchedulerPluginProcessingLatency(types.PickerPluginType, s.picker.Name(), time.Since(before))
+	if err != nil {
+		return nil, err
 	}
+	loggerDebug.Info("After running picker plugins", "result", res)
 
-	pods, err := filter.Filter(sCtx, sCtx.PodsSnapshot)
-	if err != nil || len(pods) == 0 {
-		return nil, fmt.Errorf("failed to apply filter, resulted %v pods, this should never happen: %w", len(pods), err)
+	s.runPostSchedulePlugins(sCtx, res)
+
+	return res, nil
+}
+
+func (s *Scheduler) runPreSchedulePlugins(ctx *types.Context) {
+	for _, plugin := range s.preSchedulePlugins {
+		ctx.Logger.V(logutil.DEBUG).Info("Running pre-schedule plugin", "plugin", plugin.Name())
+		before := time.Now()
+		plugin.PreSchedule(ctx)
+		metrics.RecordSchedulerPluginProcessingLatency(types.PreSchedulerPluginType, plugin.Name(), time.Since(before))
+	}
+}
+
+func (s *Scheduler) runPostSchedulePlugins(ctx *types.Context, res *types.Result) {
+	for _, plugin := range s.postSchedulePlugins {
+		ctx.Logger.V(logutil.DEBUG).Info("Running post-schedule plugin", "plugin", plugin.Name())
+		before := time.Now()
+		plugin.PostSchedule(ctx, res)
+		metrics.RecordSchedulerPluginProcessingLatency(types.PostSchedulePluginType, plugin.Name(), time.Since(before))
+	}
+}
+
+func (s *Scheduler) runFilterPlugins(ctx *types.Context) ([]types.Pod, error) {
+	loggerDebug := ctx.Logger.V(logutil.DEBUG)
+	pods := ctx.PodsSnapshot
+	loggerDebug.Info("Before running filter plugins", "pods", pods)
+	for _, filter := range s.filters {
+		loggerDebug.Info("Running filter plugin", "plugin", filter.Name())
+		before := time.Now()
+		filteredPods, err := filter.Filter(ctx, pods)
+		metrics.RecordSchedulerPluginProcessingLatency(types.FilterPluginType, filter.Name(), time.Since(before))
+		if err != nil || len(filteredPods) == 0 {
+			return nil, fmt.Errorf("failed to apply filter, resulted %v pods, this should never happen: %w", len(filteredPods), err)
+		}
+		pods = filteredPods
+		loggerDebug.Info("Filter plugin result", "plugin", filter.Name(), "pods", pods)
+	}
+	loggerDebug.Info("After running filter plugins", "pods", pods)
+	return pods, nil
+}
+
+func (s *Scheduler) runScorerPlugins(ctx *types.Context, pods []types.Pod) error {
+	loggerDebug := ctx.Logger.V(logutil.DEBUG)
+	loggerDebug.Info("Before running score plugins", "pods", pods)
+	for _, pod := range pods {
+		score, err := runScorersForPod(ctx, s.scorers, pod)
+		if err != nil {
+			return err
+		}
+		pod.SetScore(score)
+	}
+	loggerDebug.Info("After running score plugins", "pods", pods)
+	return nil
+}
+
+// Iterate through each scorer in the chain and accumulate the scores.
+func runScorersForPod(ctx *types.Context, scorers []types.Scorer, pod types.Pod) (float64, error) {
```

Comment:
this should be part of the Scheduler functions, e.g., `func (s *Scheduler) runScorersForPod` and then we shouldn't pass scorers as argument.